File: THCS/2022_2023/NgheAn_Vinh_2223/Bai2/GHH.py
def sum_divisors(x):
    total = 0
    for i in range(1, int(x**0.5) + 1):
        if x % i == 0:
            total += i
            if i != x // i:  # Tránh tính gấp đôi nếu i và x/i là giống nhau
                total += x // i
    return total

def is_nearly_perfect(x):
    divisor_sum = sum_divisors(x)
    return 2 * x <= divisor_sum

def main():
    with open("GHH.INP", "r") as input_file:
        N = int(input_file.readline().strip())
        A = list(map(int, input_file.readline().strip().split()))

    with open("GHH.OUT", "w") as output_file:
        for number in A:
            if is_nearly_perfect(number):
                output_file.write("1\n")
            else:
                output_file.write("0\n")

if __name__ == "__main__":
    main()

File: THCS/2022_2023/NgheAn_Vinh_2223/Bai2/GHH.cpp
#include <iostream>
#include <vector>
#include <fstream>
using namespace std;

// Hàm tính tổng các ước dương của số x
long long sumDivisors(int x) {
    long long sum = 0;
    for (int i = 1; i * i <= x; ++i) {
        if (x % i == 0) {
            sum += i;
            if (i != x / i) {
                sum += x / i;
            }
        }
    }
    return sum;
}

// Hàm kiểm tra số gần hoàn hảo
bool isNearlyPerfect(int x) {
    long long divisorSum = sumDivisors(x);
    return (2 * x <= divisorSum);
}

int main() {
    ifstream inputFile("GHH.INP");
    ofstream outputFile("GHH.OUT");

    int N;
    inputFile >> N; // Đọc số lượng phần tử N
    vector<int> A(N);

    // Đọc dãy số A
    for (int i = 0; i < N; ++i) {
        inputFile >> A[i];
    }

    // Kiểm tra từng phần tử trong dãy có phải là số "gần hoàn hảo" hay không
    for (int i = 0; i < N; ++i) {
        if (isNearlyPerfect(A[i])) {
            outputFile << "1" << endl;
        } else {
            outputFile << "0" << endl;
        }
    }

    // Đóng file
    inputFile.close();
    outputFile.close();

    return 0;
}
